### PR TITLE
change: increase autoqasm transpiler code coverage to 100%

### PR DESCRIPTION
**Pull request by rmshaffer**
*Issue #, if available:*
https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=32618778

*Description of changes:*
Increase code coverage to 100% for `transpiler` module of AutoQASM.
- Removes unused code, and code that apparently cannot be reached.
- Adds tests for various edge cases which are handled by the transpiler.

*Testing done:*
All existing and new tests pass.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by speller26 — APPROVED** [review]

File: src/braket/experimental/autoqasm/api/api.py [hunk 1]
```diff
@@ -53,21 +54,25 @@ def function(f: Callable) -> Callable[[Any], aq_program.Program]:
...
+    try:
+        if f.__doc__ is None:
+            f.__doc__ = ""
+        f.__doc__ += f"""
 
 Keyword Args:
     {aq_program.ProgramOptions.NUM_QUBITS.value} (int): Configuration to set the total number of
         qubits to declare in the program.
 """
+    except Exception as e:
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/api/api.py:66]: What kind of Exception?

File: src/braket/experimental/autoqasm/transpiler/transpiler.py [hunk 2]
```diff
@@ -335,35 +304,10 @@ def _try_convert_actual(
             _log_callargs(converted_f, effective_args, kwargs)
     except Exception as e:
         logging.log(1, "Error transforming entity %s", target_entity, exc_info=True)
-        if is_autograph_strict_conversion_mode():
-            raise
         exc = e
     return converted_f, exc
 
 
-def _fall_back_unconverted(
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/transpiler/transpiler.py:344]: Cool to see the deleted code!

File: test/unit_tests/braket/experimental/autoqasm/test_transpiler.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            h(q0)
+            cnot(q0, q1)
+
+    expected = """OPENQASM 3.0;
+qubit[4] __qubits__;
+h __qubits__[1];
+cnot __qubits__[1], __qubits__[3];"""
+
+    a = MyClass()
+    assert aq.function(a.bell)(1, 3).to_ir() == expected
```
Comment (laurencap) [inline test/unit_tests/braket/experimental/autoqasm/test_transpiler.py:82]: These are cool tests! It'd be neat to see how it looks with the decorators in their normal places, too.

**Review by laurencap — APPROVED** [review]

File: src/braket/experimental/autoqasm/api/api.py (hunk 1 above)
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/api/api.py:66]: AttributeError. Yes, I'll update this to be more granular. Occurs when the decorated object does not have a writable `__doc__` property. This would likely lead to another exception later (e.g., the object is not callable, etc.) but it's more informative to have that later exception thrown, rather than this one. I'll add a comment here describing this as well.

File: test/unit_tests/braket/experimental/autoqasm/test_transpiler.py (hunk 3 above)
Comment (rmshaffer) [reply test/unit_tests/braket/experimental/autoqasm/test_transpiler.py:82]: Yes, I'll add those cases too.
